### DISCUSSION TITLE
[axidraw] New brushless servo option

### DISCUSSION THREAD

**Discussion by escline** [thi-ng/umbrella#396] [Ideas]
Recently found this project as I was looking for something to use with my AxiDraw plotter. I also recently purchased their new [brushless servo upgrade kit](https://shop.evilmadscientist.com/productsmenu/953). The first thing I noticed when trying an AxiDraw example was that the pen servo was not moving up and down. I dug into it a bit and found that since the new servo uses different EBB driver board pins, it needs to be configured when using the serial API. Evil Mad Scientist [added](https://axidraw.com/doc/py_api/#api-changelog) a new option ([penlift](https://axidraw.com/doc/py_api/#penlift)) to the Python API to allow specifying the narrow-band brushless servo.

I'd like to look into this and possibly contribute this new option into the axidraw package.

**Comment by postspectacular** [comment]
Hi @escline - welcome & oh that's great to hear, someone with actual 1st hand experience of that upgrade (I'm tempted myself, but will hold back for now). I will look at the new option too and try to find out how this needs to be implemented... 👍

**Comment by escline** [comment]
I've been looking at this for the past few days now. I just wasn't sure if I needed to put in an issue or just use this discussion board.

Investigating this is what led me to find the Windows build issue, and create that PR to fix. 😀 

Last night I looked at the Python API and after enabling debugging (and inserting a logger line to output the serial commands sent), I was able to compare the servo setup commands sent for the standard servo and the new brushless servo. I used the Python _plot.py_ example.

Standard servo setup serial commands:
```
SC,8,8
SC,4,20641
SC,5,15248
SC,11,1618
SC,12,1079
...
```

Brushless servo setup:
```
SC,8,1
SC,4,9720
SC,5,7560
SC,11,231
SC,12,154
...
```

According to the [EBB Commands reference](http://evil-mad.github.io/EggBot/ebb.html#EBB_Command_Reference) :
`SC,8,<value>` - Sets the number of RC servo PWM channels. 8 (default) for standard servo. 1 for brushless/narrow-band servo.
`SC,4,<value>` - Servo min value for output position.
`SC,5,<value>` - Servo max value for output position.
`SC,11,<value>` - Rate of change for servo when going up.
`SC,12,<value>` - Rate of change for servo when going down.

I found the PenHandler [servo_setup](https://github.com/evil-mad/axidraw/blob/master/inkscape%20driver/pen_handling.py#L399) method in the Python API that does this initial setup. Looks like these serial commands setup the servo PWM channels and timings.

Also in my serial commands compare, the servo pin is specified in the `SP` (Set Pen State) command. And the brushless servo defaults are faster as well.

Standard servo pen down & up:
`SP,1,130,1` - Set to servo min position (down). 130ms duration. Pin 1
`SP,0,146,1` - Set to servo max position (up). 146ms. Pin 1

Brushless/narrow-band servo pen down & up:
`SP,1,59,2` - Set to servo min position (down). 59ms. Pin 2
`SP,0,61,2` - Set to servo max position (up). 61ms. Pin 2

Now onto the axidraw package. I've started by adding a `penlift` property to the [AxiDrawOpts](https://github.com/thi-ng/umbrella/blob/develop/packages/axidraw/src/axidraw.ts#L37) configuration object. I think we can then use this to calculate the new default delays/timings, and make any servo changes needed in the serial commands.
